### ISSUE TITLE
Is it possible to give a resolver for the underlying value of an UnionGraphType?

### ISSUE THREAD

**Issue by njlr** [graphql-dotnet/graphql-dotnet#3464]
I have a `UnionGraphType` where the underlying type is a container that may have an _A_ or a _B_. I want to be able to resolve the underlying value in a similar way to how I resolve the type. 

Underlying union class:

```csharp
public class BookOrMovie
{
    private readonly bool isBook; 
    private readonly Book book;
    private readonly Movie movie;

    public bool IsBook { get { return isBook; } }

    public Book Book { get { return book; } }

    public Movie Movie { get { return movie; } }

    public BookOrMovie(Book book)
    {
        this.isBook = true;
        this.book = book;
        this.movie = null;
    }

    public BookOrMovie(Movie movie)
    {
        this.isBook = false;
        this.book = null;
        this.movie = movie;
    }
}
```

Resolve type logic:

```csharp
      (x : obj) => {
        var bom = x as BookOrMovie;

        if (bom.IsBook) {
          return new GraphQLTypeReference("Book");
        } else {
          return new GraphQLTypeReference("Movie");
        }
      }
```

The issue is that my `Book` and `Movie` types do not understand `BookOrMovie`. 

Is there a way to resolve the value of a `UnionGraphType`?

```csharp
      (x : obj) => {
        var bom = x as BookOrMovie;

        if (bom.IsBook) {
          return bom.Book as obj;
        } else {
          return bom.Movie as obj;
        }
      }
```

**Comment by Shane32** [comment]
No but just return the underlying type directly. Then if you’re using `ObjectGraphType<Book>` and `ObjectGraphType<Movie>` then IsTypeOf will automatically be set to detect the proper object type. 

So to summarize, whatever field resolver currently returns a BookOrMovie should instead return either Book or Movie

**Comment by Shane32** [comment]
So you shouldn’t need to write the “resolve type logic” (although you can with the ResolveType delegate), and the “resolve value logic” should be embedded in the field resolver that returns the value for the union. In fact you may not need the BookOrMovie class at all

**Comment by njlr** [comment]
The problem is that this commits the underlying data-model to having an "is-a" relationship between the union type and the object type. Is this unavoidable?

**Comment by Shane32** [comment]
I guess I'm not exactly sure what you're asking.

**Comment by Shane32** [comment]
There is no place to set a resolver on a Union type AFTER its object type has been determined, but PRIOR to resolvers on the object type, as posed in your original question.

**Comment by njlr** [comment]
> I guess I'm not exactly sure what you're asking.

Sure, let me try again 🙂 

Suppose my domain is:

 - Books
 - Movies
 - Media (union of books and movies)

In C#, I can model this using an interface: 

 - `class Book : IMedia`
 - `class Movie : IMedia` 

`graphql-dotnet` works well for this arrangement. 

 - `Book` corresponds to an `ObjectGraphType`
 - `Movie` corresponds to an `ObjectGraphType`
 - `IMedia` corresponds to a `UnionGraphType`

I can implement `ResolveType` like this: 

```csharp
(x) => {
  if (x is Book) {
    return GraphQLTypeReference("Book");
  }

  if (x is Movie) {
    return GraphQLTypeReference("Movie");
  }

  throw new Exception("Unsupported type for Media union")
}
```

However, for other reasons unrelated to using GraphQL, I don't want to model my domain in this way. 

Instead, I have three sealed classes like this:

 - `class Book`
 - `class Movie`
 - `class BookOrMovie`

And

 - `Book` corresponds to an `ObjectGraphType`
 - `Movie` corresponds to an `ObjectGraphType`
 - `BookOrMovie` corresponds to a `UnionGraphType` (`Media`)

I can convert an `BookOrMovie` into either a `Book` or a `Movie` using the properties it provides. 

```csharp
var bookOrMovie = // ...

if (bookOrMovie.IsBook) {
  var book = bookOrMovie.Book;

  Console.WriteLine("The book is " + book.ToString());
}

if (bookOrMovie.IsMovie) {
  var movie = bookOrMovie.Movie;

  Console.WriteLine("The movie is " + movie.ToString());
}
```

However, how can I make this work with `graphql-dotnet`? 

When I resolve a field of type `Media`, I return a `BookOrMovie`. But this doesn't work because `BookOrMovie` cannot be cast to `Book` or `Movie`. 

It seems that I am forced to use inheritance when implementing a GraphQL union.

**Comment by Shane32** [comment]
Can you post a sample field definition that returns `BookOrMovie`?  This might be one this example:

```cs
Field<BookOrMovieType>("getById")
    .Argument<int>("id")
    .Resolve(context => {
        var id = context.GetArgument<int>("id");
        var bookOrMovie = context.RequestServices.GetRequiredService<ObjectService>().GetById(id);
        return bookOrMovie;
    });
```

**Comment by njlr** [comment]
Something like this:

```csharp
Field<List<BookOrMovieType>>("allMedia")
    .Resolve(context => {
        var results = new List<BookOrMovie>();

        results.Add(new BookOrMovie(new Movie("Top Gun")));
        results.Add(new BookOrMovie(new Book("Superforecasting")));

        return results;
    });
```

(Sorry if it does not compile, I am mentally converting from F#)

**Comment by Shane32** [comment]
```cs
Field<List<BookOrMovieType>>("allMedia")
    .Resolve(context => {
        var results = new List<object>();

        results.Add(Convert(new BookOrMovie(new Movie("Top Gun"))));
        results.Add(Convert(new BookOrMovie(new Book("Superforecasting"))));

        return results;
    });

object Convert(BookOrMovie obj)
{
    return obj is Book bookObj ? bookObj : obj is Movie movieObj ? movieObj : throw new InvalidOperationException();
}
```

**Comment by Shane32** [comment]
You shouldn't need to write a `ResolveType` method.  But you'd need to use this conversion method for anything methods that return a BookOrMovie.

**Comment by sungam3r** [comment]
> Book corresponds to an ObjectGraphType
    Movie corresponds to an ObjectGraphType
    BookOrMovie corresponds to a UnionGraphType (Media)

or BookOrMovie corresponds to a InterfaceGraphType (Media). Maybe interface is more suitable to you here.
